### PR TITLE
Clarify interoperable parsing expectations (3.0.4 port of #3732)

### PR DESCRIPTION
**Pull request by handrews**
_[**EDIT:** After much discussion, based on today's TDC call this PR is now a direct port of #3732, as the intended model had shifted from 2.0's "copy the JSON/YAML and then apply OAS Semantics" to "the reference target has OAS Semantics, but the implications of this were not entirely worked out."]_

This is sort-of a port of #3732, but modified because the scenario involved was mostly assumed to be valid based on the more direct text of OAS 2.0.  So instead of making the scenario implementation-defined, I opted to acknowledge that it's the common assumption _(there is probably better wording for this, please feel free to suggest - I fear that my current wording is too skeptical or judgemental)_, but also that the text does not cleary require it in 3.0.  And then recommend against doing it because it is not required to be supported in 3.1.  See the commit message below for more details of how this is different from the 3.1.1 PR.

----

As discovered through the OASComply project, certain referencing scenarios are ambiguous, with different authorities holding contradictory interpretations regarding whether and how they are to be supported.

This particular scenario, where the same JSON/YAML object is parsed in different semantic contexts, is well-defined if the requirements of 2.0 are assumed to apply despite no longer appearing in the text.  However, the scenario is *not* well-defined in 3.1, so we should advise against it.

While the 3.1 version of this change made the scenario explicitly implementation-defined, it seems better to acknowledge the assumption here while still noting that it is not clearly required.

This also removes the confusing reference to JSON Schema for "$ref" as the Reference Object *explicitly* states that it is governed by JSON Reference and *not* JSON Schema.  It also enumerates more of the referencing keywords like we do in 3.1.

### CODE REVIEW

**Review by ralfhandl — DISMISSED** [review]

**Review by ralfhandl — APPROVED** [review]

**Review by miqui — APPROVED** [review]

**Comment by handrews** [comment]
I've asked a few people who were around for 2.0 and 3.0 to weigh in on the wording here, so even though this has two approvals, let's not merge it just yet please (but thank you to the approvers so far!)

File: versions/3.0.4.md
Comment (DavidBiesack) [inline versions/3.0.4.md:142]: should "are parsed" be changed to "MUST BE parsed" as per [BCP 14](https://tools.ietf.org/html/bcp14) / [RFC2119](https://tools.ietf.org/html/rfc2119) / [RFC8174](https://tools.ietf.org/html/rfc8174)?

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md:150]: @DavidBiesack thanks- I agree with the improvement, but will probably wait until I can get some confirmation about whether that line is even correct before updating it.  The strange "this version's text" wording is me flailing for some way to say _"I've been told that this is what is supposed to happen by people who are either no longer involved or not currently responding with details, but I can't find anywhere in the text that justifies that position, which appears to be carried over from 2.0, despite the fact that the text in 2.0 that made that mostly clear has definitely been removed, which I would personally interpret as changing the requirements, but to what I am not sure."_

That seemed over-long, though 😅 (not to mention too much of a negative rant – it's important to me to respect past work whether I currently understand it or not.  The folks who made that decision had a reason, I just don't know what it was).  So yeah, "not required by this version's text" is my weirdly legalistic way of saying ¯\\_(ツ)_/¯ and hoping someone will see it and explain.

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md:145]: JSON Schema bundles aren't supported in 3.0, because `id`/`$id` is not supported (nor are any other keywords that would require scanning a whole document - see #3758 for how that is handled in 3.1).

These bullet points are not intended to allow anything as much as they are to describe what implementations already do.  Otherwise I'd have probably thrown in a MAY.

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md:142]: In this case I'd say no, it's not necessary, as the whole spec defines that sort of parsing.  This is a place where I'm trying to describe what tools tend to do without becoming prescriptive about how they do it.  Or getting too deep into the details of various options.

For example, some tools parse the underlying format and the OpenAPI aspects more-or-less at one (at least I assume somoene out there uses and event-based streaming parser or something of that sort).  Others do two separate passes.  Some do both passes every time they hit the reference, others do the JSON/YAML pass once and re-do the OAS part each time they hit the reference (particularly in 2.0).

So I didn't want someone to read my phrasing here and decide it mandated a specific implementation.  But I'm open to other perspectives.

File: versions/3.0.4.md
Comment (ralfhandl) [reply versions/3.0.4.md:150]: Regarding "although it is not required by this version of the OpenAPI Specification":

My impression is that it is explicitly *strongly suggested* by v3.0.x:

> **[4.7.23 Reference Object](https://spec.openapis.org/oas/v3.0.3#reference-object)**
> [...]
> The Reference Object is defined by [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03) and follows the same structure, behavior and rules.
> 
> For this specification, reference resolution is accomplished as defined by the JSON Reference specification and not by the JSON Schema specification.

And [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-4) says

> Resolution of a JSON Reference object SHOULD yield the referenced
> JSON value.  Implementations MAY choose to replace the reference with
> the referenced value.

It says _SHOULD_ in JSON Reference, so it is not _required_ to do it that way and only _strongly suggested_ 🙄

**Comment by handrews** [comment]
I have converted this to a draft just to make sure it does not get merged as (despite having two TSC requirements) there is still discussion going on.

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md:150]: @ralfhandl to be clear, I'm totally open to changing this PR to say "3.0 behaves just like 2.0", I just don't see that as supported by the text.  I'm happy to be overruled, though, as I just want this to be as clear as possible.  Right now I don't find it clear at all.

The part that was in 2.0 that was removed reads:

> The Swagger representation of the API is made of a single file. However, parts of the definitions can be split into separate files, at the discretion of the user.

So that, to me, says "it doesn't matter how anything is arranged, it is the 'dereferenced` single file that matters, and therefore there isn't any context in the target document."  The only context that matters is the reference context, because it is literally just one file in the end.  That has been removed in 3.0, which to me allows for the possibility of context in the target document, or context maintained between parsing.

But again, if the TSC says that 3.0 works the same as 2.0, I'm happy to add that clarification in 3.0.4.

Comment (ralfhandl) [reply versions/3.0.4.md:150]: > It doesn't say anything about what is done with that JSON value, and whether it makes sense to treat the _same_ referenced value as _different things_.

JSON Reference says

> Implementations MAY choose to replace the reference with
the referenced value.

That this is a "MAY" is even weirder than the "SHOULD" in the preceding sentence 🙄 

I interpret this as "copy&paste" without taking context into account.

Comment (ralfhandl) [reply versions/3.0.4.md:150]: > 3.0 works the same as 2.0

That's what all my code assumes: 3.0 is a superset of 2.0 with minor structural differences. After ignoring those differences, the building blocks are interpreted the same.

Comment (handrews) [reply versions/3.0.4.md:150]: MAY means it's completely optional, so (unlike SHOULD) you don't need a reason to ignore it.  So it definitely does not mean that you have to copy-paste.  In fact with JSON Schema draft-04, which also cites that same JSON Reference draft, you _cannot do so_ in all circumstances.  Which is actually true in OpenAPI as well, not that that stops people from trying.

But again, what is the point of debating the JSON Reference wording here?  What we need is for someone to say what the intent was and how it should be written about.  I can find implementations that assume it's the same as 2.0.  I can find implementations that have no clue what 2.0 was like and implement 3.0 differently based on the different wording.

What was the intent?  (probably the same as 2.0 but then why was the wording changed so substantially?) And how should it be described? (I have no clue and don't like what I've written)

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md:145]: @DavidBiesack let me know if I did not answer your question fully - I'm not sure if you were just talking about [formal bundling](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-bundling) or something else.

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md:150]: @ralfhandl in case it's not clear, I largely agree that your interpretation is the most likely intent.  It just doesn't connect enough in my head (given my own reading of things) for me to know what to do here.  To me, this is all in "resulted in the common interpretation" territory.  But that makes it sound like people were interpreting things that aren't there.  And honestly, that's what it looks like to me, but I know that's not what it looks like to others.

But how should it be phrased?  As a SHOULD recommendation?  As implementation-defined?  Is this a MUST that just didn't get spelled out, and if so are we comfortable enough to put it in as a MUST because implementations would have carried it over from 2.0?  I'd be comfortable with that if someone who knows the intent said so.

File: versions/3.0.4.md
Comment (RomanHotsiy) [inline versions/3.0.4.md:148]: Extra space

```suggestion
In version 2.0 of this specification, this section specified that the OAD is a "single file", which (together with the treatment of `$ref` as a URL resolved separately each time it is encountered), resulted in the common interpretation that if the same object is parsed multiple times in contexts that require it to be parsed as _different_ Object types, then as long as the reference target is valid for each Object type, then there is no error.  An example would be parsing an empty JSON object as a reference target twice: Once as a Path Item Object and once as a Schema Object. Since both Object types allow empty objects, this is syntactically valid for both Objects.
```

File: versions/3.0.4.md
Comment (RomanHotsiy) [reply versions/3.0.4.md:150]: "resulted in the common interpretation", it it's the common interpretation let's record it in the spec text.

> I know that's not what it looks like to others.

Do you have any specific examples of misinterpretation in the context of OpenAPI 3.0.x?

**Review by RomanHotsiy — APPROVED** [review]

**Comment by handrews** [comment]
@RomanHotsiy I'm not sure I know of specific misinterpretations, more like assumptions from those coming from JSON Schema and adding OpenAPI support.  They (we) tend to assume referencing works like it's intended to in 3.1, because there's nothing obvious that says otherwise in 3.0.  @ralfhandl 's reading of the JSON Reference spec is valid, but that'a level of textual analysis most people don't do.

I'm probably being overly paranoid from some past debates.  If no one else shows up with more context in tomorrow morning's call, I'll probably incorporate the feedback, take a shot at better wording, and move ahead.

**Comment by handrews** [comment]
Based on today's TDC call, the intention regarding `$ref` and documents had shifted by 3.0:

* In 2.0, first you copied the target JSON/YAML and then you applied the OAS semantics
* In 3.0, the target is assumed to have OAS semantics, however the details of how that works were never entirely worked out

Because of this, the 3.1.1 wording was deemed appropriate, with one change (the 3.0 Schema Object does not have its own `$ref`; Reference Objects are used in its place).  This is now a straight port of 3.1.1 plus a commit to remove the Schema Object `$ref` clause.

File: versions/3.0.4.md
Comment (ralfhandl) [inline versions/3.0.4.md:149]: What if a referenced document also contains an OpenAPI object, for example if one API references a type of another API, like here:

- [People](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/oasis-tcs/odata-openapi/main/examples/People.openapi3.json)
- [Products](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/oasis-tcs/odata-openapi/main/examples/Products.openapi3.json)

Would this perfectly working use case now be disallowed?

**Review by ralfhandl — CHANGES_REQUESTED** [review]
The new text is slightly too restrictive, can't do that (again) in a patch version.

**Comment by handrews** [comment]
@ralfhandl 

> The new text is slightly too restrictive, can't do that in a patch version.

Which text do you mean?  If it's the text I did not change, then it's already in and that is not a reason to block this PR (although it is a good reason to open a new PR to address the separate problem).

**Comment by handrews** [comment]
@ralfhandl I have filed PR #3819 for the "entry document" concern - let's please get this PR resolved and handle that separately!

**Comment by handrews** [comment]
@ralfhandl I have filed PR #3820 (replacing 3819 which had errors and a misnamed branch) for the "entry document" concern.  Let's please get this PR merged and I will resolve conflicts in that one and we can finish that discussion separately.

File: versions/3.0.4.md
Comment (handrews) [reply versions/3.0.4.md]: All of this language got taken out so I'm marking this resolved.

**Review by miqui — APPROVED** [review]

**Comment by miqui** [comment]
I am ok with the language of this PR.  Any other concern should be handled in a new PR, etc..etc.  @handrews  has already done that with #3820

**Review by ralfhandl — APPROVED** [review]

**Pull request merged**
